File: js/src/domains/compositeLayerEditor/types/compositeLayer.ts
import { FlumeCommentMap } from 'flume';
import {
    MetricType,
    MetricValue,
    ScaleDomainRange,
} from '../../dataLayers/types/metrics';
import { FlumeGraph } from './flumeGraph';

export type CompositeLayer = {
    id: number;
    name: string;
    graph: FlumeGraph;
    /** Flume canvas comments (annotations), persisted alongside the graph. */
    comments: FlumeCommentMap;
    metric_type: number | null;
    /** Full resulting layer, returned on save so it can be shown on the map immediately. */
    metric_type_detail: MetricType | null;
    /** Requested legend, mirrored on the graph's output node (auto/reference/linear/…). */
    legend_type: string;
    /** Manually-configured buckets, set for a concrete legend type. */
    legend_config: ScaleDomainRange;
    legend_reference_metric_type: number | null;
    created_at: string;
    updated_at: string;
};

export type CompositeLayerListItem = Pick<
    CompositeLayer,
    | 'id'
    | 'name'
    | 'metric_type'
    | 'legend_type'
    | 'legend_config'
    | 'created_at'
    | 'updated_at'
>;

export type SaveCompositeLayerPayload = {
    /** When provided, updates the existing composite layer instead of creating one. */
    id?: number;
    /** Sent by the node editor; a metadata-only save leaves it out. */
    graph?: FlumeGraph;
    comments?: FlumeCommentMap;
    name?: string;
    category?: string;
    description?: string;
    units?: string;
    unit_symbol?: string;
    is_population?: boolean;
    legend_type?: string;
    legend_config?: { domain: (number | string)[]; range: string[] };
};

/** Result of evaluating a graph without persisting it, shaped for the map component. */
export type CompositePreview = {
    units: string;
    unit_symbol: string;
    legend_type: string;
    legend_config: { domain: number[] | string[]; range: string[] };
    metric_values: MetricValue[];
    /** Distinct non-null years present in `metric_values`, newest first (empty for a timeless graph). */
    years: number[];
};

export type CompositePreviewState = {
    status: 'idle' | 'loading' | 'error' | 'ready';
    data: CompositePreview | null;
    error?: string | null;
};

File: models/composite_layer.py
from django.conf import settings
from django.db import models


class CompositeLayer(models.Model):
    """
    A composite data layer built with the visual node editor.

    Stores the serialized Flume graph so it can be re-opened and edited, plus a link to the
    ``MetricType`` generated when the graph is executed. Running the graph produces a new
    ``MetricType`` (category "Composite") and its ``MetricValue`` rows, so a composite layer
    behaves like any other data layer on the map and can itself be used as a graph input.
    """

    class LegendType(models.TextChoices):
        # ``MetricType.LegendType`` plus two modes resolved when the graph runs: "auto" computes the
        # buckets from the values, "reference" copies another layer's legend.
        AUTO = "auto"
        REFERENCE = "reference"
        LINEAR = "linear"
        THRESHOLD = "threshold"
        ORDINAL = "ordinal"

    class Meta:
        app_label = "snt_malaria"
        ordering = ["-updated_at"]

    account = models.ForeignKey("iaso.Account", on_delete=models.CASCADE, related_name="composite_layers")
    name = models.CharField(max_length=255)
    # Serialized Flume graph: { nodeId: { id, type, x, y, inputData, connections } }.
    graph = models.JSONField(default=dict)
    # Flume canvas comments (annotations), kept separate from the graph so the evaluator, which
    # iterates the graph as a node map, never sees them: { commentId: { id, text, x, y, ... } }.
    comments = models.JSONField(default=dict, blank=True)
    # The MetricType produced by the last successful run of this graph.
    metric_type = models.ForeignKey(
        "iaso.MetricType",
        null=True,
        blank=True,
        on_delete=models.SET_NULL,
        related_name="composite_layer",
    )
    # Legend requested for this layer, mirrored on the graph's output node. `MetricType.legend_type`
    # and `legend_config` hold what a run resolved it to.
    legend_type = models.CharField(max_length=40, choices=LegendType.choices, default=LegendType.AUTO)
    # Manually configured {domain, range}, set for a concrete legend type (empty otherwise).
    legend_config = models.JSONField(default=dict, blank=True)
    # Layer whose legend is copied when legend_type is "reference"; unset means the first one
    # connected to the output.
    legend_reference_metric_type = models.ForeignKey(
        "iaso.MetricType",
        null=True,
        blank=True,
        on_delete=models.SET_NULL,
        related_name="+",
    )
    created_by = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        null=True,
        blank=True,
        on_delete=models.SET_NULL,
        related_name="+",
    )
    created_at = models.DateTimeField(auto_now_add=True)
    updated_at = models.DateTimeField(auto_now=True)

    def __str__(self):
        return self.name
